File: Chapter1/one-eight.py
# if there is any row or column in the matrix with 0, make the whole row/column 0
def zero_matrix(matrix):
    col = False
    row = False
    for y in range(0, len(matrix)):
        if (matrix[y][0] == 0):
            col = True
    for x in range(0, len(matrix[0])):
        if (matrix[0][x] == 0):
            row = True
    for y in range(1, len(matrix)):
        for x in range(1, len(matrix[y])):
            if (matrix[y][x] == 0):
                if (x != 0 ):
                    matrix[0][x] = 0
                if (y != 0):
                    matrix[y][0] = 0
    # makes the columns 0
    for item in range(1,len(matrix[0])):
        if (matrix[0][item] == 0):
            for y in range(len(matrix)):
                matrix[y][item] = 0
    # makes the rows 0
    for item in range(1, len(matrix)):
        if (matrix[item][0] == 0):
            for x in range(len(matrix[0])):
                matrix[item][x] = 0
    #check if first row and column should be 0
    if (row):
        for x in range(len(matrix[0])):
            matrix[0][x] = 0
    if (col):
        for y in range(len(matrix)):
            matrix[y][0] = 0
    return matrix
    
def print_matrix(matrix):
    for y in matrix:
        print(y)

File: CrackingTheCodingInterview/Chapter1/OneEight.java
package CrackingTheCodingInterview.Chapter1;

// if a there is a 0 in a row or col, changes that whole row/col to 0 in place
public class OneEight 
{

    static int[][] zero(int[][] matrix)
    {
        boolean col = false;
        boolean row = false;
        for (int x : matrix[0])
        {
            if (x == 0)
            {
                row = true;
            }
        }
        for (int y = 0; y < matrix.length; y++)
        {
            if (y == 0)
            {
                col = true;
            }
        }
        for (int x = 1; x < matrix.length; x++)
        {
            for (int y = 1; y < matrix[0].length; y++)
            {
                if(matrix[x][y] == 0)
                {
                    matrix[x][0] = 0;
                    matrix[0][y] = 0;
                }
            }
        }
        for (int i = 1; i < matrix[0].length; i++)
        {
            if (matrix[0][i] == 0)
            {
                for (int y = 1; y < matrix.length; y++)
                {
                    matrix[y][i] = 0;
                }
            }
        }
        for (int i = 1; i < matrix.length; i++)
        {
            if (matrix[i][0] == 0)
            {
                for (int x = 1; x < matrix[0].length; x++)
                {
                    matrix[i][x] = 0;
                }
            }
        }
        if(row)
        {
            for (int i = 0; i < matrix[0].length; i++)
            {
                matrix[0][i] = 0;
            }
        }
        if(col)
        {
            for (int i = 0; i < matrix.length; i++)
            {
                matrix[i][0] = 0;
            }
        }
        return matrix;
    }
    
}